File: Matecska/Models/Scoring.swift
import Foundation

/// Egy válasz pontszámának összetevői.
struct ScoreBreakdown: Equatable {
    let base: Int
    let bonus: Int
    let penalty: Int

    var total: Int { base + bonus - penalty }

    static let zero = ScoreBreakdown(base: 0, bonus: 0, penalty: 0)
}

/// A sebességbónusz időzítése: eddig teljes a bónusz, és ekkortól nincs már.
struct BonusTiming: Equatable {
    let fullBonusUntil: TimeInterval
    let noBonusAfter: TimeInterval

    /// Összeadás, kivonás: 3 másodpercig teljes, 10 másodpercre fogy el.
    static let quick = BonusTiming(fullBonusUntil: 3, noBonusAfter: 10)
    /// Szorzás, osztás: türelmesebb, 5 másodpercig teljes, 15 másodpercre fogy el.
    static let patient = BonusTiming(fullBonusUntil: 5, noBonusAfter: 15)
}

/// Pontozási szabályok.
/// Helyes válasz: alap pont + sebességbónusz, ami a művelet időzítése szerint
/// egy ideig maximális, majd lineárisan csökken nullára.
/// Helytelen válasz: levonás (az alap pont fele).
enum Scoring {
    static let basePoints = 10
    static let maxBonus = 10
    static let wrongAnswerPenalty = 5

    static func points(correct: Bool, elapsed: TimeInterval, timing: BonusTiming) -> ScoreBreakdown {
        guard correct else {
            return ScoreBreakdown(base: 0, bonus: 0, penalty: wrongAnswerPenalty)
        }
        return ScoreBreakdown(base: basePoints, bonus: bonus(elapsed: elapsed, timing: timing), penalty: 0)
    }

    /// A még elérhető bónusz aránya 0…1 között; a bónuszcsík ezt rajzolja.
    static func bonusFraction(elapsed: TimeInterval, timing: BonusTiming) -> Double {
        if elapsed <= timing.fullBonusUntil { return 1 }
        if elapsed >= timing.noBonusAfter { return 0 }
        return 1 - (elapsed - timing.fullBonusUntil) / (timing.noBonusAfter - timing.fullBonusUntil)
    }

    static func bonus(elapsed: TimeInterval, timing: BonusTiming) -> Int {
        Int((Double(maxBonus) * bonusFraction(elapsed: elapsed, timing: timing)).rounded())
    }
}

File: web/src/core/scoring.ts
/** Egy válasz pontszámának összetevői. */
export interface ScoreBreakdown {
  base: number;
  bonus: number;
  penalty: number;
}

export function totalPoints(score: ScoreBreakdown): number {
  return score.base + score.bonus - score.penalty;
}

/**
 * A sebességbónusz időzítése.
 * `fullBonusUntil` másodpercig a teljes bónusz jár, utána minden `decayIntervalMs`
 * milliszekundumos periódusban pontosan eggyel csökken, nulláig.
 * A periódus később szintek szerint változhat.
 */
export interface BonusTiming {
  fullBonusUntil: number;
  decayIntervalMs: number;
}

export const DEFAULT_DECAY_INTERVAL_MS = 1000;

/** Egyjegyű, fejben számolós feladatok: 1 másodperc türelmi idő, utána másodpercenként eggyel kevesebb. */
export const INSTANT_TIMING: BonusTiming = { fullBonusUntil: 1, decayIntervalMs: DEFAULT_DECAY_INTERVAL_MS };
/** Írásbeli összeadás, kivonás: 3 másodpercig teljes, utána másodpercenként eggyel kevesebb. */
export const QUICK_TIMING: BonusTiming = { fullBonusUntil: 3, decayIntervalMs: DEFAULT_DECAY_INTERVAL_MS };
/** Írásbeli szorzás, osztás: türelmesebb, 5 másodpercig teljes, utána másodpercenként eggyel kevesebb. */
export const PATIENT_TIMING: BonusTiming = { fullBonusUntil: 5, decayIntervalMs: DEFAULT_DECAY_INTERVAL_MS };

/** Alappont a háromjegyű típusoknál; a többi típus a saját `basePoints` értékét adja át. */
export const BASE_POINTS = 10;
/** A bónusz maximuma a típus alappontja; ez a háromjegyű alap. */
export const MAX_BONUS = BASE_POINTS;
export const WRONG_ANSWER_PENALTY = 1;
/** Vegyes (operandust is kérdező) gyakorlásnál a bónusz ennyiszer lassabban fogy. */
export const MIXED_SLOWDOWN = 1.5;

/** Az időzítés lassítása: a fogyás periódusa `factor`-szorosára nő. */
export function scaledTiming(timing: BonusTiming, factor: number): BonusTiming {
  return { ...timing, decayIntervalMs: timing.decayIntervalMs * factor };
}

/** A még járó bónusz pontban: `max` (a típus alappontja), majd periódusonként eggyel kevesebb. */
export function bonus(elapsed: number, timing: BonusTiming, max: number = MAX_BONUS): number {
  const overtimeMs = Math.max(0, elapsed - timing.fullBonusUntil) * 1000;
  const lostPoints = Math.floor(overtimeMs / timing.decayIntervalMs);
  return Math.max(0, max - lostPoints);
}

/** A még elérhető bónusz aránya 0…1 között; a bónuszcsík ezt rajzolja. */
export function bonusFraction(elapsed: number, timing: BonusTiming, max: number = MAX_BONUS): number {
  return bonus(elapsed, timing, max) / max;
}

/**
 * Helyes válasz: a típus alappontja + ugyanannyi maximumú sebességbónusz.
 * Helytelen válasz: 1 pont levonás.
 */
export function points(correct: boolean, elapsed: number, timing: BonusTiming, base: number = BASE_POINTS): ScoreBreakdown {
  if (!correct) return { base: 0, bonus: 0, penalty: WRONG_ANSWER_PENALTY };
  return { base, bonus: bonus(elapsed, timing, base), penalty: 0 };
}
